### PR TITLE
Remove type-based name resolution from the AST 

### PR DESCRIPTION
**Pull request by wingedtrashcan**
The main purpose of this PR is to replace `VarF` and `VarB` with a single constructor `Var` and move the type information to `DeclVariable` like it's usually done in programming languages.

The PR also does some other things, see below.

### CODE REVIEW

File: src/Nightfall/Lang/Internal/Types.hs [hunk 1]
```diff
@@ -87,7 +95,7 @@ data Expr_ =
 -- | Simple, internal type, not exposed
 data Statement_ =
     -- | Variable declaration
-      DeclVariable VarName Expr_  -- ^ let a = 634
+      DeclVariable VarType VarName Expr_  -- ^ let a = 634
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Internal/Types.hs:98]: We now annotate the declaration site of a variable with its type.

File: src/Nightfall/Lang/Internal/Types.hs [hunk 2]
```diff
@@ -96,6 +104,7 @@ data Statement_ =
 
     -- | Conditionals
     | IfElse Expr_ [Statement_] [Statement_] -- ^ condition if-block else-block
+    | Repeat Natural [Statement_]
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Internal/Types.hs:107]: I'll need that for bigints.

File: src/Nightfall/Lang/Internal/Types.hs [hunk 3]
```diff
@@ -39,41 +40,48 @@ type VarName = String
...
-    | Bo Bool   -- ^ true/false, 1/0
+      Literal Literal
 
     | UnOp UnOp Expr_
     | BinOp BinOp Expr_ Expr_
 
-    -- | Variables
-    | VarF VarName     -- ^ "calling" a variable of type Felt by its name (e.g. "foo")
-    | VarB VarName     -- ^ same, but with boolean variable
+    | Var VarName
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Internal/Types.hs:84]: And we don't annotate usage sites of variables anymore.

File: src/Nightfall/Lang/Types.hs [hunk 4]
```diff
@@ -51,24 +49,36 @@ module Nightfall.Lang.Types ( Felt
...
 import Data.Word (Word32)
+import qualified Data.Map.Strict as Map
 
 -- | Expression wrapper type, typed for safety, exposed to use
 newtype Expr a = Expr
     { unExpr :: Expr_
     } deriving (Eq, Show)
 
+class IsBuiltin a where
+    literal :: a -> Literal
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Types.hs:65]: Just some convenience.

File: src/Nightfall/MASM/Types.hs [hunk 5]
```diff
@@ -86,7 +90,9 @@ data Instruction
   | LocStore MemoryIndex -- loc_store.i
   | LocLoad MemoryIndex -- loc_load.i
   | MemLoad (Maybe MemoryIndex) -- mem_load[.i]
+  | MemLoadw (Maybe MemoryIndex) -- mem_loadw[.i]
   | MemStore (Maybe MemoryIndex) -- mem_store[.i]
+  | MemStorew (Maybe MemoryIndex) -- mem_storew[.i]
```
Comment (wingedtrashcan) [inline src/Nightfall/MASM/Types.hs:95]: Added a bunch of other stuff I'm gonna need for bigints.

File: src/Nightfall/Targets/Miden.hs [hunk 6]
```diff
@@ -279,45 +299,39 @@ transpileBinOp LowerEq        = [ MASM.Lte ]
...
 transpileExpr (GetAt arr i) = do
     vars <- gets variables
-    -- TODO: also check that @arr@ is an array.
-    case unMemoryIndex <$> Map.lookup arr vars of
+    case Map.lookup arr vars of
         Nothing -> error $ "Array variable \"" ++ arr ++ "\" unknown (undeclared)"
-        Just arrPos -> do
+        Just varInfo -> do
+            unless (isArrayOfFelt $ _varInfoType varInfo) $
+                error $ "'" ++ arr ++ "' is used as an array, but it's not one"
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:333]: Added a check that `arr` is an array here and in the clause for `SetAt`.

**Review by allezlom13 — APPROVED** [review]

**Pull request merged**